### PR TITLE
PLFM-8655: Project storage usage cache and limits

### PR DESCRIPTION
**Pull request by marcomarasca**

### CODE REVIEW

File: lib/jdomodels/src/main/resources/schema/limits/ProjectStorageLimit-ddl.sql [hunk 1]
```diff
@@ -0,0 +1,13 @@
+CREATE TABLE `PROJECT_STORAGE_LIMIT` (
+  `ID` BIGINT NOT NULL,
+  `ETAG` CHAR(36) NOT NULL,
+  `CREATED_BY` BIGINT NOT NULL,
+  `CREATED_ON` TIMESTAMP NOT NULL,
+  `MODIFIED_BY` BIGINT NOT NULL,
+  `MODIFIED_ON` TIMESTAMP NOT NULL,
+  `PROJECT_ID` BIGINT NOT NULL,
+  `STORAGE_LOCATION_ID` BIGINT NOT NULL,
+  `MAX_BYTES` BIGINT NOT NULL,
```
Comment (john-hill) [inline lib/jdomodels/src/main/resources/schema/limits/ProjectStorageLimit-ddl.sql]: Need a script to one-time set all the limits (unlimited).  Need to settle on a way to encode no limit.  Easy filter to find what is limited and unlimited.

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/limits/ProjectStorageLimitsDao.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	 */
+	ProjectStorageLocationLimit setStorageLocationLimit(long userId, ProjectStorageLocationLimit limit);
+	
+	/**
+	 * 
+	 * @param projectId
+	 * @param storageLocationId
+	 * @return The limit for the given project and storage location ids pair
+	 */
+	Optional<ProjectStorageLocationLimit> getStorageLocationLimit(Long projectId, Long storageLocationId);
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/limits/ProjectStorageLimitsDao.java:49]: this should be an error if nothing is there.

File: lib/jdomodels/src/main/resources/schema/limits/ProjectStorageData-ddl.sql [hunk 3]
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE `PROJECT_STORAGE_DATA` (
+  `PROJECT_ID` BIGINT NOT NULL,
+  `ETAG` CHAR(36) NOT NULL,
+  `MODIFIED_ON` TIMESTAMP NOT NULL,
+  `RUNTIME_MS` BIGINT NOT NULL,
+  `STORAGE_LOCATION_DATA` JSON NOT NULL,
+  PRIMARY KEY (`PROJECT_ID`) 
+);
```
Comment (john-hill) [inline lib/jdomodels/src/main/resources/schema/limits/ProjectStorageData-ddl.sql:9]: cascade delete on the project.

File: lib/jdomodels/src/main/resources/schema/limits/ProjectStorageLimit-ddl.sql [hunk 4]
```diff
@@ -0,0 +1,13 @@
+CREATE TABLE `PROJECT_STORAGE_LIMIT` (
+  `ID` BIGINT NOT NULL,
+  `ETAG` CHAR(36) NOT NULL,
+  `CREATED_BY` BIGINT NOT NULL,
```
Comment (john-hill) [inline lib/jdomodels/src/main/resources/schema/limits/ProjectStorageLimit-ddl.sql:4]: should ref user.

File: lib/jdomodels/src/main/resources/schema/limits/ProjectStorageLimit-ddl.sql [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+  `CREATED_ON` TIMESTAMP NOT NULL,
+  `MODIFIED_BY` BIGINT NOT NULL,
+  `MODIFIED_ON` TIMESTAMP NOT NULL,
+  `PROJECT_ID` BIGINT NOT NULL,
+  `STORAGE_LOCATION_ID` BIGINT NOT NULL,
```
Comment (john-hill) [inline lib/jdomodels/src/main/resources/schema/limits/ProjectStorageLimit-ddl.sql:9]: cascade on the storage location.

File: lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/TableIndexDAOImplTest.java [hunk 6]
```diff
@@ -168,7 +169,7 @@ public void after() {
 		if (tableId != null && tableIndexDAO != null) {
 			tableIndexDAO.deleteTable(tableId);
 		}
-		tableIndexDAO.truncateIndex();
+		//tableIndexDAO.truncateIndex();
```
Comment (john-hill) [inline lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/TableIndexDAOImplTest.java]: uncomment

**Review by john-hill — APPROVED** [review]

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/limits/ProjectStorageLimitManager.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+			.setLocations(locations.values().stream().sorted(LOCATION_COMPARATOR).collect(Collectors.toList()));
+	}
+	
+	@WriteTransaction
+	public void refreshProjectStorageData(Long projectId) {
+		if (storageUsageDao.isStorageDataModifiedOnAfter(projectId, clock.now().toInstant().minus(CACHE_UPDATE_FREQUENCY))) {
+			return;
+		}
+		
+		ProjectStorageData data = replicationDao.computeProjectStorageData(projectId);
```
Comment (marcomarasca) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/limits/ProjectStorageLimitManager.java:110]: Note: there is an additional complication: a user can create a S3StorageLocationSetting that points to the default synapse bucket. This storage location will have a different id than the default storage location so this grouping won't work if we want a limit on our bucket per project.

**Review by john-hill — APPROVED** [review]

**Pull request merged**